src/main/kotlin/org/rust/cargo/project/settings/RustProjectSettingsService.kt

```diff
@@ -5,53 +5,80 @@
 
 package org.rust.cargo.project.settings
 
-import com.intellij.openapi.Disposable
+import com.intellij.openapi.components.Storage
+import com.intellij.openapi.components.StoragePathMacros
 import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiManager
 import com.intellij.util.ThreeState
 import com.intellij.util.io.systemIndependentPath
-import com.intellij.util.messages.Topic
 import com.intellij.util.xmlb.annotations.Transient
-import org.jetbrains.annotations.TestOnly
+import org.rust.cargo.project.configurable.RsProjectConfigurable
+import org.rust.cargo.project.settings.RustProjectSettingsService.State
 import org.rust.cargo.toolchain.ExternalLinter
 import org.rust.cargo.toolchain.RsToolchain
 import org.rust.cargo.toolchain.RsToolchainBase
 import org.rust.cargo.toolchain.RsToolchainProvider
 import org.rust.openapiext.isUnitTestMode
+import org.rust.openapiext.showSettingsDialog
 import java.nio.file.Paths
-import kotlin.reflect.KProperty1
-import kotlin.reflect.full.findAnnotation
-import kotlin.reflect.full.memberProperties
 
-interface RustProjectSettingsService {
+val Project.rustSettings: RustProjectSettingsService
+    get() = getService(RustProjectSettingsService::class.java)
+        ?: error("Failed to get ${serviceName}Service for $this")
 
-    data class State(
-        var version: Int? = null,
+val Project.toolchain: RsToolchainBase?
+    get() {
+        val toolchain = rustSettings.state.toolchain
+        return when {
+            toolchain != null -> toolchain
+            isUnitTestMode -> RsToolchainBase.suggest()
+            else -> null
+        }
+    }
+
+private const val serviceName: String = "RustProjectSettings"
+
+@com.intellij.openapi.components.State(name = serviceName, storages = [
+    Storage(StoragePathMacros.WORKSPACE_FILE),
+    Storage("misc.xml", deprecated = true)
+])
+class RustProjectSettingsService(project: Project) : RsProjectSettingsServiceBase<State>(project, State()) {
+    val toolchain: RsToolchainBase? get() = state.toolchain
+    val explicitPathToStdlib: String? get() = state.explicitPathToStdlib
+    val autoShowErrorsInEditor: ThreeState get() = ThreeState.fromBoolean(state.autoShowErrorsInEditor)
+    val autoUpdateEnabled: Boolean get() = state.autoUpdateEnabled
+    val compileAllTargets: Boolean get() = state.compileAllTargets
+    val useOffline: Boolean get() = state.useOffline
+    val macroExpansionEngine: MacroExpansionEngine get() = state.macroExpansionEngine
+    val doctestInjectionEnabled: Boolean get() = state.doctestInjectionEnabled
+
+    class State : StateBase<State>() {
         @AffectsCargoMetadata
-        var toolchainHomeDirectory: String? = null,
-        var autoShowErrorsInEditor: Boolean = true,
-        var autoUpdateEnabled: Boolean = true,
+        var toolchainHomeDirectory by string()
+        var autoShowErrorsInEditor by property(true)
+        var autoUpdateEnabled by property(true)
```

Comment:
Did you check that xml layout with using new API is the same and migration is not needed?

Reply:
Yes, if you open an old project settings are loaded properly.